// @assister/rasa/dist/rasa/ion-refresher-content.entry.js
import { r as registerInstance, c as config, d as getIonMode, h, H as Host } from './chunk-292824f0.js';
import { s as sanitizeDOMString } from './chunk-bb9ec3bf.js';

class RefresherContent {
    constructor(hostRef) {
        registerInstance(this, hostRef);
    }
    componentWillLoad() {
        if (this.pullingIcon === undefined) {
            this.pullingIcon = config.get('refreshingIcon', 'arrow-down');
        }
        if (this.refreshingSpinner === undefined) {
            const mode = getIonMode(this);
            this.refreshingSpinner = config.get('refreshingSpinner', config.get('spinner', mode === 'ios' ? 'lines' : 'crescent'));
        }
    }
    hostData() {
        const mode = getIonMode(this);
        return {
            class: {
                [mode]: true,
            }
        };
    }
    __stencil_render() {
        return [
            h("div", { class: "refresher-pulling" }, this.pullingIcon &&
                h("div", { class: "refresher-pulling-icon" }, h("ion-icon", { icon: this.pullingIcon, lazy: false })), this.pullingText &&
                h("div", { class: "refresher-pulling-text", innerHTML: sanitizeDOMString(this.pullingText) })),
            h("div", { class: "refresher-refreshing" }, this.refreshingSpinner &&
                h("div", { class: "refresher-refreshing-icon" }, h("ion-spinner", { name: this.refreshingSpinner })), this.refreshingText &&
                h("div", { class: "refresher-refreshing-text", innerHTML: sanitizeDOMString(this.refreshingText) }))
        ];
    }
    render() { return h(Host, this.hostData(), this.__stencil_render()); }
}

export { RefresherContent as ion_refresher_content };
