// @ionic/core/dist/collection/components/menu-controller/menu-controller.js
import { Build } from "@stencil/core";
import { menuController } from '../../utils/menu-controller';
export class MenuController {
    constructor() {
        if (Build.isDev) {
            console.warn(`[DEPRECATED][ion-menu-controller] Use the menuController export from @ionic/core:
  import { menuController } from '@ionic/core';`);
        }
    }
    /**
     * Open the menu. If a menu is not provided then it will open the first
     * menu found. If the specified menu is `start` or `end`, then it will open
     * the enabled menu on that side. Otherwise, it will try to find the menu
     * using the menu's `id` property. If a menu is not found then it will
     * return `false`.
     *
     * @param menu The menuId or side of the menu to open.
     */
    open(menu) {
        return menuController.open(menu);
    }
    /**
     * Close the menu. If a menu is specified, it will close that menu.
     * If no menu is specified, then it will close any menu that is open.
     * If it does not find any open menus, it will return `false`.
     *
     * @param menu The menuId or side of the menu to close.
     */
    close(menu) {
        return menuController.close(menu);
    }
    /**
     * Toggle the menu open or closed. If the menu is already open, it will try to
     * close the menu, otherwise it will try to open it. Returns `false` if
     * a menu is not found.
     *
     * @param menu The menuId or side of the menu to toggle.
     */
    toggle(menu) {
        return menuController.toggle(menu);
    }
    /**
     * Enable or disable a menu. Disabling a menu will not allow gestures
     * for that menu or any calls to open it. This is useful when there are
     * multiple menus on the same side and only one of them should be allowed
     * to open. Enabling a menu will automatically disable all other menus
     * on that side.
     *
     * @param enable If `true`, the menu should be enabled.
     * @param menu The menuId or side of the menu to enable or disable.
     */
    enable(enable, menu) {
        return menuController.enable(enable, menu);
    }
    /**
     * Enable or disable the ability to swipe open the menu.
     *
     * @param enable If `true`, the menu swipe gesture should be enabled.
     * @param menu The menuId or side of the menu to enable or disable the swipe gesture on.
     */
    swipeGesture(enable, menu) {
        return menuController.swipeGesture(enable, menu);
    }
    /**
     * Get whether or not the menu is open. Returns `true` if the specified
     * menu is open. If a menu is not specified, it will return `true` if
     * any menu is currently open.
     *
     * @param menu The menuId or side of the menu that is being checked.
     */
    isOpen(menu) {
        return menuController.isOpen(menu);
    }
    /**
     * Get whether or not the menu is enabled. Returns `true` if the
     * specified menu is enabled. Returns `false` if a menu is disabled
     * or not found.
     *
     * @param menu The menuId or side of the menu that is being checked.
     */
    isEnabled(menu) {
        return menuController.isEnabled(menu);
    }
    /**
     * Get a menu instance. If a menu is not provided then it will return the first
     * menu found. If the specified menu is `start` or `end`, then it will return the
     * enabled menu on that side. Otherwise, it will try to find the menu using the menu's
     * `id` property. If a menu is not found then it will return `null`.
     *
     * @param menu The menuId or side of the menu.
     */
    get(menu) {
        return menuController.get(menu);
    }
    /**
     * Get the instance of the opened menu. Returns `null` if a menu is not found.
     */
    getOpen() {
        return menuController.getOpen();
    }
    /**
     * Get all menu instances.
     */
    getMenus() {
        return menuController.getMenus();
    }
    /**
     * Get whether or not a menu is animating. Returns `true` if any
     * menu is currently animating.
     */
    isAnimating() {
        return menuController.isAnimating();
    }
    /**
     * Registers a new animation that can be used with any `ion-menu` by
     * passing the name of the animation in its `type` property.
     *
     * @param name The name of the animation to register.
     * @param animation The animation function to register.
     */
    async registerAnimation(name, animation) {
        return menuController.registerAnimation(name, animation);
    }
    static get is() { return "ion-menu-controller"; }
    static get originalStyleUrls() { return {
        "$": ["menu-controller.scss"]
    }; }
    static get styleUrls() { return {
        "$": ["menu-controller.css"]
    }; }
    static get methods() { return {
        "open": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<boolean>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu to open.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu to open."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Open the menu. If a menu is not provided then it will open the first\nmenu found. If the specified menu is `start` or `end`, then it will open\nthe enabled menu on that side. Otherwise, it will try to find the menu\nusing the menu's `id` property. If a menu is not found then it will\nreturn `false`.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu to open."
                    }]
            }
        },
        "close": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<boolean>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu to close.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu to close."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Close the menu. If a menu is specified, it will close that menu.\nIf no menu is specified, then it will close any menu that is open.\nIf it does not find any open menus, it will return `false`.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu to close."
                    }]
            }
        },
        "toggle": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<boolean>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu to toggle.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu to toggle."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Toggle the menu open or closed. If the menu is already open, it will try to\nclose the menu, otherwise it will try to open it. Returns `false` if\na menu is not found.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu to toggle."
                    }]
            }
        },
        "enable": {
            "complexType": {
                "signature": "(enable: boolean, menu?: string | null | undefined) => Promise<HTMLIonMenuElement | undefined>",
                "parameters": [{
                        "tags": [{
                                "text": "enable If `true`, the menu should be enabled.",
                                "name": "param"
                            }],
                        "text": "If `true`, the menu should be enabled."
                    }, {
                        "tags": [{
                                "text": "menu The menuId or side of the menu to enable or disable.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu to enable or disable."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLIonMenuElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLIonMenuElement | undefined>"
            },
            "docs": {
                "text": "Enable or disable a menu. Disabling a menu will not allow gestures\nfor that menu or any calls to open it. This is useful when there are\nmultiple menus on the same side and only one of them should be allowed\nto open. Enabling a menu will automatically disable all other menus\non that side.",
                "tags": [{
                        "name": "param",
                        "text": "enable If `true`, the menu should be enabled."
                    }, {
                        "name": "param",
                        "text": "menu The menuId or side of the menu to enable or disable."
                    }]
            }
        },
        "swipeGesture": {
            "complexType": {
                "signature": "(enable: boolean, menu?: string | null | undefined) => Promise<HTMLIonMenuElement | undefined>",
                "parameters": [{
                        "tags": [{
                                "text": "enable If `true`, the menu swipe gesture should be enabled.",
                                "name": "param"
                            }],
                        "text": "If `true`, the menu swipe gesture should be enabled."
                    }, {
                        "tags": [{
                                "text": "menu The menuId or side of the menu to enable or disable the swipe gesture on.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu to enable or disable the swipe gesture on."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLIonMenuElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLIonMenuElement | undefined>"
            },
            "docs": {
                "text": "Enable or disable the ability to swipe open the menu.",
                "tags": [{
                        "name": "param",
                        "text": "enable If `true`, the menu swipe gesture should be enabled."
                    }, {
                        "name": "param",
                        "text": "menu The menuId or side of the menu to enable or disable the swipe gesture on."
                    }]
            }
        },
        "isOpen": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<boolean>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu that is being checked.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu that is being checked."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Get whether or not the menu is open. Returns `true` if the specified\nmenu is open. If a menu is not specified, it will return `true` if\nany menu is currently open.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu that is being checked."
                    }]
            }
        },
        "isEnabled": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<boolean>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu that is being checked.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu that is being checked."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Get whether or not the menu is enabled. Returns `true` if the\nspecified menu is enabled. Returns `false` if a menu is disabled\nor not found.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu that is being checked."
                    }]
            }
        },
        "get": {
            "complexType": {
                "signature": "(menu?: string | null | undefined) => Promise<HTMLIonMenuElement | undefined>",
                "parameters": [{
                        "tags": [{
                                "text": "menu The menuId or side of the menu.",
                                "name": "param"
                            }],
                        "text": "The menuId or side of the menu."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLIonMenuElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLIonMenuElement | undefined>"
            },
            "docs": {
                "text": "Get a menu instance. If a menu is not provided then it will return the first\nmenu found. If the specified menu is `start` or `end`, then it will return the\nenabled menu on that side. Otherwise, it will try to find the menu using the menu's\n`id` property. If a menu is not found then it will return `null`.",
                "tags": [{
                        "name": "param",
                        "text": "menu The menuId or side of the menu."
                    }]
            }
        },
        "getOpen": {
            "complexType": {
                "signature": "() => Promise<HTMLIonMenuElement | undefined>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLIonMenuElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLIonMenuElement | undefined>"
            },
            "docs": {
                "text": "Get the instance of the opened menu. Returns `null` if a menu is not found.",
                "tags": []
            }
        },
        "getMenus": {
            "complexType": {
                "signature": "() => Promise<HTMLIonMenuElement[]>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "HTMLIonMenuElement": {
                        "location": "global"
                    }
                },
                "return": "Promise<HTMLIonMenuElement[]>"
            },
            "docs": {
                "text": "Get all menu instances.",
                "tags": []
            }
        },
        "isAnimating": {
            "complexType": {
                "signature": "() => Promise<boolean>",
                "parameters": [],
                "references": {
                    "Promise": {
                        "location": "global"
                    }
                },
                "return": "Promise<boolean>"
            },
            "docs": {
                "text": "Get whether or not a menu is animating. Returns `true` if any\nmenu is currently animating.",
                "tags": []
            }
        },
        "registerAnimation": {
            "complexType": {
                "signature": "(name: string, animation: AnimationBuilder | ((menu: MenuI) => IonicAnimation)) => Promise<void>",
                "parameters": [{
                        "tags": [{
                                "text": "name The name of the animation to register.",
                                "name": "param"
                            }],
                        "text": "The name of the animation to register."
                    }, {
                        "tags": [{
                                "text": "animation The animation function to register.",
                                "name": "param"
                            }],
                        "text": "The animation function to register."
                    }],
                "references": {
                    "Promise": {
                        "location": "global"
                    },
                    "AnimationBuilder": {
                        "location": "import",
                        "path": "../../interface"
                    },
                    "MenuI": {
                        "location": "import",
                        "path": "../../interface"
                    },
                    "IonicAnimation": {
                        "location": "import",
                        "path": "../../interface"
                    }
                },
                "return": "Promise<void>"
            },
            "docs": {
                "text": "Registers a new animation that can be used with any `ion-menu` by\npassing the name of the animation in its `type` property.",
                "tags": [{
                        "name": "param",
                        "text": "name The name of the animation to register."
                    }, {
                        "name": "param",
                        "text": "animation The animation function to register."
                    }]
            }
        }
    }; }
}
